Add-Devise-authentication,-Users-model,-Add-tests

fh0enix opened the pull request:
Title: Adding Devise Authentication and Users Model with Test Cases

Description:

This pull request introduces essential changes to enhance the security and user management of our application. Here's a breakdown of what this pull request accomplishes:

Devise Authentication Integration: Devise authentication has been integrated into our application. This ensures a robust and reliable user authentication system, providing a secure gateway for our users to access the platform.
![image](https://github.com/fh0enix/books_reader/assets/13628239/20e9afda-8f76-43da-986a-a4ac02baa50c)

Users Model Implementation: A dedicated "User" model has been introduced. This model encapsulates essential user-related data and functionalities, paving the way for efficient user management.
![image](https://github.com/fh0enix/books_reader/assets/13628239/6c565b84-ec53-4bd7-b671-380bcfaa32f6)

Comprehensive Test Suite: To maintain the reliability of our application, an extensive suite of tests has been included. These tests rigorously examine the functionality of the authentication system, ensuring that our users' data remains protected and our features operate as expected.
![image](https://github.com/fh0enix/books_reader/assets/13628239/b0e58e2d-2ce8-464c-abac-323cd19952d2)


The integration of Devise and the introduction of the Users model significantly improve our application's security and user management. The comprehensive test coverage guarantees the stability of these changes.

Please review these modifications thoroughly, paying special attention to the security aspects and the impact on user functionality. Your feedback and suggestions are highly appreciated to ensure the successful implementation of this critical update.

Reviewer:

@loqimean

fh0enix committed:
✨ feat:

add device gem

fh0enix committed:
✨ feat:

add user model

fh0enix committed:
✨ feat:

generate devise:views

fh0enix committed:
✨ feat:

add validates for  first_name, last_name

fh0enix committed:
🐞 fix:

correct validation

fh0enix committed:
🌈 style:

user log

fh0enix committed:
🌈 style:

add

fh0enix committed:
🧪 test:

test

fh0enix committed:
🦄 refactor:

del unnecessary view

fh0enix committed:
🧪 test:

user_spec test

loqimean commented on app/controllers/application_controller.rb:
```diff
@@ -1,3 +1,11 @@
-
 class ApplicationController < ActionController::Base
+  before_action :configure_permitted_parameters, if: :devise_controller?
+  #before_action :authenticate_user!
```
видаляй зайве

fh0enix replied:
solved

loqimean commented on app/controllers/application_controller.rb:
```diff
@@ -1,3 +1,11 @@
+  before_action :configure_permitted_parameters, if: :devise_controller?
+  #before_action :authenticate_user!
+
+protected
+
+def configure_permitted_parameters
+  devise_parameter_sanitizer.permit(:sign_up, keys:[:first_name, :last_name, :user_name])
+  devise_parameter_sanitizer.permit(:account_update, keys: [:first_name, :last_name, :user_name])
```
де відступи? кодстайл

fh0enix replied:
solved

loqimean commented on app/models/user.rb:
```diff
@@ -0,0 +1,8 @@
+  devise :database_authenticatable, :registerable,
+         :recoverable, :rememberable, :validatable
+
+validates :first_name, :last_name, :user_name, presence: true
```
де відступи?

fh0enix replied:
sorry
solved

loqimean commented on app/views/layouts/_header.html.erb:
```diff
@@ -0,0 +1,31 @@
+			</li>
+		</ul>
+		<a rel="noopener noreferrer" href="/books" aria-label="Back to homepage" class="flex items-center p-2 duration-75 delay-1000 animate-pulse">
+			<svg xmlns="http://www.w3.org/2000/svg" fill="currentColor" viewBox="0 0 32 32" class="w-8 h-8 dark:text-cyan-400">
```
`inline_svg`

fh0enix replied:
solved

loqimean commented on app/views/layouts/application.html.erb:
```diff
@@ -12,11 +12,14 @@
       <%= render 'layouts/alerts' %>
     </div>
 
+
```
```suggestion
```

fh0enix replied:
solved

loqimean commented on app/views/layouts/application.html.erb:
```diff
@@ -12,11 +12,14 @@
 
   <body class="bg-gray-400">
+    <%= render 'layouts/header' %>
+
```
```suggestion
```

fh0enix replied:
solved

loqimean replied:
не солвед)

fh0enix replied:
солвед)

loqimean commented on app/views/layouts/_header.html.erb:
```diff
@@ -0,0 +1,31 @@
+				<%= link_to "Sign in", user_session_path, class: "self-center px-8 py-3 rounded" %>
+				<%= link_to "Sign Up", new_user_registration_path, class: "self-center px-8 py-3 font-semibold rounded dark:bg-cyan-400 dark:text-gray-900" %>
+			<% end %>
+		</div>
+		<button class="p-4 lg:hidden">
+			<svg xmlns="http://www.w3.org/2000/svg" fill="none" viewBox="0 0 24 24" stroke="currentColor" class="w-6 h-6 dark:text-gray-100">
+				<path stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M4 6h16M4 12h16M4 18h16"></path>
+			</svg>
+		</button>
+	</div>
+</header>
```
чому тут по 4 пробіли? якщо не більше. виправляй

fh0enix replied:
The code містить пробіли, які відповідають HTML-відступам для кращої читабельності коду

loqimean commented on app/views/layouts/_header.html.erb:
```diff
@@ -0,0 +1,31 @@
+			<% end %>
+		</div>
+		<button class="p-4 lg:hidden">
+			<svg xmlns="http://www.w3.org/2000/svg" fill="none" viewBox="0 0 24 24" stroke="currentColor" class="w-6 h-6 dark:text-gray-100">
```
`inline_svg`

fh0enix replied:
solved

loqimean commented on config/routes.rb:
```diff
@@ -1,12 +1,8 @@
-
 Rails.application.routes.draw do
+  devise_for :users
```
тобі прям всі роути тре?

fh0enix replied:
solved

loqimean commented on db/seeds.rb:
```diff
@@ -1,3 +1,3 @@
 require "factory_bot_rails"
 
-FactoryBot.create_list(:book, 100)
+FactoryBot.create_list(:book, 5)
```
а юзери?

fh0enix replied:
solved

loqimean commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,19 @@
+
+RSpec.describe User, type: :model do
+  describe 'validations' do
+    it { should validate_presence_of(:email) }
```
це старий синтаксис. міняй

fh0enix replied:
solved

loqimean commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,19 @@
+    it { should validate_presence_of(:user_name) }
+  end
+
+  describe 'factory' do
+    it 'has a valid factory' do
+      expect(FactoryBot.create(:user)).to be_valid
+    end
+  end
```
це вверх і без класу юзай

fh0enix replied:
deleted it

loqimean commented on spec/rails_helper.rb:
```diff
@@ -35,6 +34,7 @@
 RSpec.configure do |config|
   # Remove this line if you"re not using ActiveRecord or ActiveRecord fixtures
   config.fixture_path = Rails.root.join("spec/fixtures").to_s
+  config.include Devise::Test::ControllerHelpers, type: :controller
```
не в той тип передав

fh0enix replied:
solved?

loqimean commented on spec/requests/users_spec.rb:
```diff
@@ -0,0 +1,50 @@
+    context 'with valid attributes' do
+      it 'creates a new user' do
+        expect do
+          post user_registration_path, params: { user: FactoryBot.attributes_for(:user) }
```
фекторі_бот клас дефайниться глобально, нема сенсу писати константу

fh0enix replied:
solved

loqimean replied:
не солвед)

fh0enix replied:
now I have desided it

fh0enix committed:
🐞 fix:

fix warning from expert

fh0enix committed:
✨ feat:

add gem Inline SVG and change _header

loqimean commented on config/routes.rb:
```diff
@@ -1,12 +1,8 @@
   get "up" => "rails/health#show", as: :rails_health_check
 
-  # Defines the root path route ("/")
   root "books#index"
```
на верх перед девайсом

fh0enix replied:
solved

loqimean commented on config/routes.rb:
```diff
@@ -1,12 +1,8 @@
-
 Rails.application.routes.draw do
+  devise_for :users, only: [:sessions, :registrations]
```
```suggestion
  devise_for :users, only: [:sessions, :registrations]

```

fh0enix replied:
solved

loqimean commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,11 @@
+require 'rails_helper'
+
+RSpec.describe User, type: :model do
+    it { is_expected.to validate_presence_of(:email) }
+    it { is_expected.to validate_uniqueness_of(:email).case_insensitive }
+    it { is_expected.to validate_presence_of(:password) }
+    it { is_expected.to validate_length_of(:password).is_at_least(6) }
+    it { is_expected.to validate_presence_of(:first_name) }
+    it { is_expected.to validate_presence_of(:last_name) }
+    it { is_expected.to validate_presence_of(:user_name) }
```
два пробіли

fh0enix replied:
solved

loqimean commented on app/controllers/application_controller.rb:
```diff
@@ -1,3 +1,11 @@
 class ApplicationController < ActionController::Base
+  before_action :configure_permitted_parameters, if: :devise_controller?
+
+protected
+
+def configure_permitted_parameters
+  devise_parameter_sanitizer.permit(:sign_up, keys:[:first_name, :last_name, :user_name])
+
+  devise_parameter_sanitizer.permit(:account_update, keys: [:first_name, :last_name, :user_name])
+end
```
два пробіли

fh0enix replied:
solved

loqimean commented on app/views/layouts/_header.html.erb:
```diff
@@ -0,0 +1,28 @@
+<nav class="sticky top-0 p-4 dark:bg-gray-800 dark:text-gray-100 rounded-b-2xl">
+	<div class="container flex justify-between h-16 mx-auto">
+		<a href="/books" class="flex items-center p-2">
+			<%= inline_svg "book.svg", class: "w-8 h-8 text-cyan-400" %>
+		</a>
+		<ul class="items-stretch hidden lg:flex">
+			<li class="flex">
+				<% if current_user %>
+					<a><%= link_to "UPLOAD NEW BOOK", new_book_path, class: "duration-75 delay-1000 animate-pulse flex items-center px-4 -mb-1 border-b-2 dark:border-transparent dark:text-cyan-400 dark:border-cyan-400" %></a>
+				<% end %>
+			</li>
+		</ul>
+		<div class="items-center flex-shrink-0 hidden lg:flex">
+			<% if current_user %>
```
код стайл

fh0enix replied:
solved

fh0enix committed:
The diff includes changes to the CSS file, controller file, header file, and routes file.

In the CSS file, a new class called `.btn-hello` is created. It has the properties of duration, delay, and animation. The `.btn-green` class is also updated to have a white background and a gray border.

In the controller file, the `configure_permitted_parameters` method is updated to allow for the parameters `:first_name`, `:last_name`, and `:user_name` to be permitted for both sign up and account update.

In the header file, the navigation bar is updated to include a new link to the `new_book_path`. The link is only visible to logged-in users.

In the routes file, the `root` path is updated to point to the `books#index` action.

fh0enix committed:
Update routes.rb, user_spec.rb, and users_spec.rb

fh0enix committed:
Update styles for books index and show pages

fh0enix committed:
Update styles for books index, forms, devise registrations and sessions, and book factory

fh0enix committed:
Update the background color of the login and registration pages from #f0f0f0 to #707070

fh0enix committed:
removes devise_parameter_sanitizer.permit(:account_update, keys: [:first_name, :last_name, :user_...

fh0enix committed:
```
This commit changes the primary abstract class of the ApplicationRecord model from ActiveRecord::Base to nil.

fh0enix committed:
Update books migration and add users migration

loqimean commented on app/views/books/_show_book.html.erb:
```diff
@@ -0,0 +1,23 @@
+<div class="max-w-lg p-4 shadow-md dark:dark:bg-gray-900 dark:dark:text-gray-100 rounded-2xl">
+	<div class="flex justify-between pb-4 border-bottom">
+		<div class="flex items-center">
+      <a><%= link_to "BACK TO BOOKS", books_path, class: "mb-0 capitalize dark:dark:text-cyan-400" %></a>
+		</div>
+		<a rel="noopener noreferrer" href="#">Read book</a>
+	</div>
+	<div class="space-y-4">
+		<div class="space-y-2">
+			<img src="https://source.unsplash.com/random/480x360/?4" alt="" class="book-cover">
+			<div class="flex items-center text-xs">
+				<span>ISBN:<%= book.isbn %></span>
+			</div>
+		</div>
+		<div class="space-y-2">
+			<a rel="noopener noreferrer" href="#" class="block">
+				<h4 class="text-xl font-semibold dark:dark:text-cyan-400"><%= book.author %></h4>
+				<h3 class="text-xl font-semibold dark:dark:text-gray-100"><%= book.title %></h3>
```
ну що це за приколи з табами по 8 пробілів?))

loqimean commented on app/views/layouts/_header.html.erb:
```diff
@@ -0,0 +1,25 @@
+<nav class="sticky top-0 p-4 dark:bg-gray-900 dark:text-gray-100 rounded-b-2xl">
+	<div class="container flex justify-between h-16 mx-auto">
+		<a href="/books" class="flex items-center p-2">
+			<%= inline_svg "book.svg", class: "w-8 h-8 text-cyan-400" %>
+		</a>
+		<div class="items-center flex-shrink-0 hidden lg:flex">
+			<% if current_user %>
+				<div class="self-center px-8 py-3 rounded">
+					<ul class="items-stretch hidden lg:flex">
+						<li class="flex">
+								<a><%= link_to "Hello #{current_user.first_name}! UPLOAD NEW BOOK", new_book_path, class: "btn-hello" %></a>
+						</li>
+					</ul>
+				</div>
+				<%= button_to "Log Out", destroy_user_session_path, method: :delete, class: "self-center px-8 py-3 font-semibold rounded dark:bg-amber-400 dark:text-gray-900" %>
```
та ж історія

loqimean commented on config/routes.rb:
```diff
@@ -1,12 +1,9 @@
-
 Rails.application.routes.draw do
+  root "books#index"
+
+  devise_for :users, only: [:sessions, :registrations]
```
```suggestion
  devise_for :users, only: [:sessions, :registrations]

  root "books#index"
```

fh0enix replied:
solved

fh0enix committed:
Update the book show page and header

fh0enix committed:
The navigation bar has been simplified. The "Log Out" button has been moved to the right of the navigation bar. The "Sign in" and "Sign Up" buttons have been moved to the left of the navigation bar.

fh0enix committed:
Change tab

fh0enix committed:
Update the book show page layout

fh0enix committed:
The diff shows that the root path was changed from pointing to the `books#index` action to the `rails/health#show` action.

The `devise_for` line was also modified to only include the `:sessions` and `:registrations` actions.

fh0enix committed:
The diff removes the devise_for :users, only: [:sessions, :registrations] line, which was used to enable user authentication.

loqimean commented on app/views/layouts/_header.html.erb:
```diff
@@ -0,0 +1,25 @@
+      <%= inline_svg "book.svg", class: "w-8 h-8 text-cyan-400" %>
+    </a>
+    <div class="items-center flex-shrink-0 hidden lg:flex">
+      <% if current_user %>
```
```suggestion
      <% if current_user.present? %>
```

fh0enix replied:
solved

fh0enix committed:
🌈 style:

Removed 3 instances of unnecessary whitespace.

fh0enix committed:
🌈 style:

 Update application.html.erb

fh0enix committed:
🦄 refactor:

 Updates conditional statement to check if current_user is present instead of truthy
